core/ktor/src/main/java/ir/composenews/ktor/ResponseTransformer.kt

```diff
@@ -0,0 +1,96 @@
+
+package ir.composenews.ktor
+
+import kotlin.contracts.ExperimentalContracts
```

Comment:
Nice! Do you get inspiration from somewhere else or any article?

Reply:
When I was researching API management in Ktor, I came across this library: [https://github.com/skydoves/sandwich](https://github.com/skydoves/sandwich)
Reading through its code was incredibly educational, and I felt we could draw inspiration from parts of this repo and customize it to fit our specific needs. While we could have added the entire library, my goal was to tailor it to our particular project and learn from it in the process. The Sandwich library is indeed very powerful, and customizing it for our use case not only served our project's needs but also provided a great learning opportunity. It allowed us to understand the intricacies of API response handling in Ktor while adapting the concepts to our specific requirements

Reply:
Thank you for your innovation 🙏 
Can you describe the benefit of this methods?

Reply:
@VahidGarousi 

Certainly
These extension functions provide a powerful and flexible way to handle different states of the `ApiResponse`

Chaining capability:
Most of these functions return ApiResponse<T>, allowing you to chain multiple operations. This leads to more readable and concise code.

Suspend function support:
The 'suspend' versions of these functions allow for easy integration with coroutines

Mapping functionality:
The 'suspendMap' functions allow you to transform the content of Success, Error, or Exception responses, providing flexibility in how you process the data.

Let us have the advantages of these helper methods with an example:
Using `suspendOnSuccess`: Let's look at this
```
override suspend fun syncMarketList() {
    api.getMarkets(
        "usd",
        "market_cap_desc",
        20,
        1,
        false,
    ).suspendOnSuccess {
        val toRemoteMarketDto = data.map {
            it.toRemoteMarketDto()
        }
        dao.upsertMarket(toRemoteMarketDto)
    }.onError {
        Log.d("debug", message)
    }.onException {
        Log.d("debug", message.toString())
    }
}
```
Here, `suspendOnSuccess` is used because inside the success block, you're calling another suspend function: `dao.upsertMarket(toRemoteMarketDto)`. This is exactly the "Suspend function support" you mentioned.

> This approach allows you to perform other asynchronous operations if the API call is successful, without worrying about managing threads or callbacks.


Using `suspendMap`: Now, let's look at your second example:
```
override fun fetchChart(id: String): Flow<Resource<Chart, Errors>> = flow {
    val chart = api.getMarketChart(id, "usd", 1)
    chart.suspendOnSuccess {
        suspendMap {
            emit(Resource.Success(it.data.toChart()))
        }
    }.suspendOnError {
        suspendMap {
            emit(
                Resource.Error(
                    error = Errors.ApiError(
                        it.statusCode.mapMessageStatusCode(),
                        it.statusCode.code,
                    ),
                ),
            )
        }
    }.suspendOnException {
        suspendMap {
            emit(Resource.Error(Errors.ExceptionError(it.message, throwable)))
        }
    }
}
```
Here, `suspendMap` acts as a mapper, This method allows you to transform the API result as desired and then `emit` it in the `Flow`.

> The use of `suspendMap` in this context is particularly useful when you need to perform some asynchronous operations or transformations on your data before `emitting it to the Flow`. It seamlessly integrates with the coroutine-based approach, maintaining the asynchronous nature of your operations while providing a clean and expressive way to handle different response states.


Indeed, that's exactly why we have both `suspendOnSuccess` and `onSuccess` versions, as well as corresponding versions for error and exception handling. This flexibility allows us to handle different scenarios efficiently. Let's break it down:

1. Regular versions (`onSuccess`, `onError`, `onException`):
These are used when you don't need to perform any suspend operations in the callback. They're suitable for simple transformations or actions that don't require coroutine context.

2. Suspend versions (`suspendOnSuccess`, `suspendOnError`, `suspendOnException`):
These are used when you need to perform suspend operations within the callback. They're perfect for scenarios where you need to call other suspend functions or perform long-running operations without blocking the main thread.

3. `suspendMap` for each of these:
The suspendMap function acts as a transformer for each of these callbacks. It allows you to perform asynchronous transformations on your data before finalizing the result.

**Here's a conceptual overview of how these functions relate:**
```
// Regular versions
onSuccess { /* non-suspend operations */ }
onError { /* non-suspend error handling */ }
onException { /* non-suspend exception handling */ }

// Suspend versions
suspendOnSuccess { /* can include suspend operations */ }
suspendOnError { /* can include suspend operations for error handling */ }
suspendOnException { /* can include suspend operations for exception handling */ }

// suspendMap versions for transformation
suspendOnSuccess {
    suspendMap { /* asynchronous transformation */ }
}
suspendOnError {
    suspendMap { /* asynchronous error transformation */ }
}
suspendOnException {
    suspendMap { /* asynchronous exception transformation */ }
}
```

**In conclusion,
The Sandwich library is truly excellent and provides a comprehensive explanation of its functionality. Its design and implementation are really impressive. My goal in not directly adding the library, but rather referencing it and implementing our own version, was a smart strategy for several reasons:**

**By implementing it ourselves, we're forced to dive deep into the code and understand the underlying mechanisms. This approach ensures that we're not just using a tool, but truly understanding how it works. If we need to add or modify functionality in the future, we already have a solid understanding of the codebase. This makes it easier to extend or customize the functionality to suit our specific needs.**

**This approach encourages continuous learning and improvement. By regularly revisiting and potentially expanding on this implementation, we're keeping our skills sharp and up-to-date. By implementing it ourselves, we have full control and understanding of every part of our codebase, which can be crucial for debugging and optimization.**

**While it would have been easier to simply add the library directly, this approach of "forced engagement" with the code is beneficial. It reflects our commitment to learning and mastery of our craft.**

Thank you for taking the time to read through this detailed explanation🙏